File: internal/domain/scope.go
package domain

import (
	"fmt"
	"strings"
)

// Scope is a structured scope identity (kind + key).
type Scope struct {
	Kind ScopeKind
	Key  string
}

// NewScope validates and returns a scope. Keys are trimmed.
func NewScope(kind ScopeKind, key string) (Scope, error) {
	trimmed := strings.TrimSpace(key)
	if trimmed == "" {
		return Scope{}, validationError("scope key must be non-empty")
	}
	if len(trimmed) > MaxScopeKeyLength {
		return Scope{}, validationError(
			fmt.Sprintf("scope key must be at most %d characters", MaxScopeKeyLength),
		)
	}
	return Scope{Kind: kind, Key: trimmed}, nil
}

File: src/memlore/domain/models/scope.py
from __future__ import annotations

from dataclasses import dataclass

from memlore.domain.exceptions import ValidationError
from memlore.domain.models.enums import ScopeKind

MAX_SCOPE_KEY_LENGTH = 512


@dataclass(frozen=True, slots=True)
class Scope:
    kind: ScopeKind
    key: str

    def __post_init__(self) -> None:
        key = self.key.strip()
        if not key:
            raise ValidationError("scope key must be non-empty")
        if len(key) > MAX_SCOPE_KEY_LENGTH:
            raise ValidationError(
                f"scope key must be at most {MAX_SCOPE_KEY_LENGTH} characters"
            )
        object.__setattr__(self, "key", key)
